ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -132,32 +128,71 @@ private void internalTerminalBlockReached(Bytes32 executionBlockHash) {
                     new IllegalStateException(
                         "Failed to retrieve execution payload hash from beacon block root"));
 
-    final Bytes32 lastSentForkChoiceHead =
-        lastSentForkChoiceState.map(ForkChoiceState::getHeadBlockHash).orElse(Bytes32.ZERO);
-
-    if (lastSentForkChoiceHead.equals(parentExecutionHash)) {
-      // Payload ID builds on the correct execution block
-      if (lastPayloadId.isPresent() || parentExecutionHash.isZero()) {
-        return lastPayloadId;
+    final boolean lastForkChoiceStateCorrectlyBuildsOnTop =
+        lastSentForkChoiceState
+            .map(ForkChoiceState::getHeadBlockHash)
+            .map(
+                fcsHead -> {
+                  if (fcsHead.equals(parentExecutionHash)) {
+                    return true;
+                  }
+                  return executionPayloadTerminalBlockHash.isPresent()
+                      && fcsHead.equals(executionPayloadTerminalBlockHash.get());
+                })
+            .orElse(false);
+
+    if (lastForkChoiceStateCorrectlyBuildsOnTop) {
+      // current Future Payload ID builds on the correct block
+
+      if (lastFuturePayloadId.isEmpty()) {
+        // consistency check
+        throw new IllegalStateException(
```

Comment:
nit: because of the way this is called I don't think it matters but generally methods that return a `SafeFuture` should return a `failedFuture` rather than throwing exceptions.